File: 2690-house-robber-iv/2690-house-robber-iv.java
class Solution {
    int n;
    public boolean helper(int[] nums, int k, int mid) {
        int last = -2;
        for (int i = 0; i < n; i++) {
            if (nums[i] <= mid && last < i-1) {
                k--;
                last = i;
            }
        }
        return k <= 0;
    }

    public int minCapability(int[] nums, int k) {
        int l = nums[0], r = nums[0];
        n = nums.length;
        for (int num : nums) {
            l = Math.min(l, num);
            r = Math.max(r, num);
        }

        while (l <= r) {
            int mid = l + (r - l) / 2;
            if (helper(nums, k, mid)) {
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        }

        return l;
    }
}

File: 2690-house-robber-iv/2690-house-robber-iv.cpp
class Solution {
public:
    int help(vector<int>& nums, int mid, int k) {

        int count = 0;
        for (int i = 0; i < nums.size(); i++) {
            if (nums[i] <= mid) {
                count++;
                i++;
            }
        }
        return count >= k;
    }

    int minCapability(vector<int>& nums, int k) {
        int low = nums[0];
        int high = nums[0];

        for (int i = 0; i < nums.size(); i++) {
            low = min(nums[i], low);
            high = max(nums[i], high);
        }

        while (low <= high) {
            int mid = (low + high) / 2;
            if (help(nums, mid, k)) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return low;
    }
};